In [31]:
import numpy as np
import pandas as pd
import pulp
import itertools
import gmaps
import googlemaps
import matplotlib.pyplot as plt
from urllib.parse import urlencode
import requests


API_KEY = 'AIzaSyB-u8dReCLDNbCeQXhkJ4WydLgwbHGVaAg'
gmaps.configure(api_key=API_KEY)
googlemaps = googlemaps.Client(key=API_KEY)

In [2]:
coppel = pd.DataFrame(pd.read_excel("Entregas_22_Actualizadas.xlsx"))

In [3]:
df = coppel[(coppel.fechaenrutada == "2022-01-04") & (coppel.Cedis == "Guadalajara")]

In [4]:
df

Bodega        Cedis  factura  foliocontrol  tienda  \
19766   30007  Guadalajara   623230       7222827     800   
19803   30007  Guadalajara   624114       7222844     800   
19814   30007  Guadalajara   621388       7222831     800   
19824   30007  Guadalajara   328622       7222842     396   
19870   30007  Guadalajara   328613       7222839     396   
...       ...          ...      ...           ...     ...   
31189   30007  Guadalajara   591610       7222825     800   
31192   30007  Guadalajara     2745       7222844    1150   
31197   30007  Guadalajara   625743       7222819     800   
31216   30007  Guadalajara    16048       7222830      86   
31221   30007  Guadalajara   580507       7222837     800   

                     Unnamed: 5 fechaventa fechaenrutada  ruta  codigo  ...  \
19766  30007_623230_7222827_800 2022-01-03    2022-01-04     2  150223  ...   
19803  30007_624114_7222844_800 2022-01-03    2022-01-04    36  361313  ...   
19814  30007_621388_7222831_800 2022-01-03    2022-01-04     4  468365  ...   
19824  30007_328622_7222842_396 2022-01-03    2022-01-04    36  318663  ...   
19870  30007_328613_7222839_396 2022-01-03    2022-01-04    36  343013  ...   
...                         ...        ...           ...   ...     ...  ...   
31189  30007_591610_7222825_800 2021-12-29    2022-01-04     1  361976  ...   
31192   30007_2745_7222844_1150 2022-01-03    2022-01-04    36  444033  ...   
31197  30007_625743_7222819_800 2022-01-03    2022-01-04     1  636886  ...   
31216    30007_16048_7222830_86 2021-12-29    2022-01-04     4  550663  ...   
31221  30007_580507_7222837_800 2021-12-28    2022-01-04    34  476716  ...   

       tipodeservicio num_ciudad  nombreciudad numerozona  \
19766              VB         66   GUADALAJARA        762   
19803              VB         69       ZAPOPAN        704   
19814              VB         69       ZAPOPAN        756   
19824              VB         69       ZAPOPAN        336   
19870              VB         69       ZAPOPAN        302   
...               ...        ...           ...        ...   
31189              VB         66   GUADALAJARA        920   
31192              DV         69       ZAPOPAN        746   
31197              VB         66   GUADALAJARA        780   
31216              RE         66   GUADALAJARA        814   
31221              VB        269       TEQUILA        267   

                  nombrezona numerodecasa  departamento num_codigopostal  \
19766        LOMAS DEL GALLO       1338.0             0            44760   
19803   COLINAS D LA PRIMAVE        147.0             0            45066   
19814            BUGAMBILIAS       4952.0             0            45238   
19824             VALLE REAL        685.0            27            45019   
19870   HOGARES D NVO MEXICO         36.0             H            45138   
...                      ...          ...           ...              ...   
31189        LAZARO CARDENAS       4310.0             0            44980   
31192          LA HACIENDITA       1615.0           NaN            45230   
31197              U H ATLAS       2045.0             5            44870   
31216                 DR ATL       5352.0           NaN            44399   
31221  BELLAVISTA  MAGDALENA        113.0             0            46470   

           DCF            Calle  
19766  18404.0   FERNANDO SOLIS  
19803  30204.0          CASCADA  
19814  49255.0  PASEO DE LA VIC  
19824  31008.0  PASEO SAN ARTUR  
19870  30104.0      AV AVIACION  
...        ...              ...  
31189  32208.0  MIGUEL OROZCO C  
31192  44007.0     VILLA JUAREZ  
31197  66540.0        RIO AMECA  
31216  50330.0   EZEQUIEL A DUE  
31221  46504.0    JUAN PABLO LL  

[493 rows x 21 columns]

In [5]:
df1 = df.copy()

In [6]:
df1.drop(df1[df1["foliocontrol"] == 7222830].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222831].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222832].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222833].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222834].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222835].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222837].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222838].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222839].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222840].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222841].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222842].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222843].index, inplace = True)
df1.drop(df1[df1["foliocontrol"] == 7222844].index, inplace = True)

In [7]:
zmg = []
for i in range(len(df1)):
    dom = f"{int(df1['numerodecasa'].iloc[i])},{df1['Calle'].iloc[i]},{df1['nombrezona'].iloc[i]},\
    {df1['num_codigopostal'].iloc[i]},{df1['nombreciudad'].iloc[i]}, JALISCO, MEXICO"
    zmg.append(dom)

In [8]:
zmg

['1338,FERNANDO SOLIS,LOMAS DEL GALLO,    44760,GUADALAJARA, JALISCO, MEXICO',
 '819,CANTERA,SAN MARCOS,    44330,GUADALAJARA, JALISCO, MEXICO',
 '228,COMBATE DE SACR,JARDINES DE STA ISAB,    44300,GUADALAJARA, JALISCO, MEXICO',
 '1738,CIPRES,DEL FRESNO,    44900,GUADALAJARA, JALISCO, MEXICO',
 '2151,REYES CATOLICOS,JARDINES DE LA CRUZ,    44950,GUADALAJARA, JALISCO, MEXICO',
 '428,FELIX BERNARDEL,SAN JUAN BOSCO,    44730,GUADALAJARA, JALISCO, MEXICO',
 '5422,C CENTAURO DEL,U H EL COLLY C T M,    45056,ZAPOPAN, JALISCO, MEXICO',
 '1366,FLORESTA,RINCONADA DEL BOSQUE,    44510,GUADALAJARA, JALISCO, MEXICO',
 '1066,RIO PAPALOAPAN,QUINTA VELARDE,    44430,GUADALAJARA, JALISCO, MEXICO',
 '5422,C CENTAURO DEL,U H EL COLLY C T M,    45056,ZAPOPAN, JALISCO, MEXICO',
 '1579,RIO PURIFICACIO,LAS AGUILAS,    45080,ZAPOPAN, JALISCO, MEXICO',
 '228,COMBATE DE SACR,JARDINES DE STA ISAB,    44300,GUADALAJARA, JALISCO, MEXICO',
 '1030,JUAN ANTONIO DE,LA NORMAL,    44260,GUADALAJARA, JALISCO, MEXICO',
 

In [9]:
def extract_lat_long(address, data_type = 'json'):
    endpoint = f"https://maps.googleapis.com/maps/api/geocode/{data_type}"
    params = {"address": address, "key": API_KEY}
    url_params = urlencode(params)
    url = f"{endpoint}?{url_params}"
    r = requests.get(url)
    if r.status_code not in range(200,299):
       return {}
    latlng = {}
    try:
        latlng = r.json()['results'][0]['geometry']['location']
    except:
        pass
    return latlng.get("lat"), latlng.get("lng")

In [10]:
def extractor(folio, lista):
    for i in range(len(folio)):
        dom = extract_lat_long(folio[i])
        lista.append(dom)

In [11]:
koordinaten = []
extractor(zmg, koordinaten)

In [12]:
koordinaten

[(20.6794392, -103.2698917),
 (20.6966845, -103.3149996),
 (20.7160767, -103.3150292),
 (20.6579742, -103.3660875),
 (20.6478777, -103.3782135),
 (20.6792166, -103.3146951),
 (20.6485868, -103.4258915),
 (20.6525832, -103.3872271),
 (20.6568445, -103.3301345),
 (20.6485868, -103.4258915),
 (20.6254755, -103.412596),
 (20.7160767, -103.3150292),
 (20.6788253, -103.3590221),
 (20.7101873, -103.3425569),
 (20.6154324, -103.3495323),
 (20.6978123, -103.2784999),
 (20.6622997, -103.2733636),
 (20.6315324, -103.3713304),
 (20.6700916, -103.273583),
 (20.6485868, -103.4258915),
 (20.6138285, -103.3485143),
 (20.6284358, -103.3815293),
 (20.6794019, -103.274831),
 (20.7169418, -103.3319842),
 (20.7160767, -103.3150292),
 (20.7210185, -103.3579529),
 (20.6362018, -103.4113772),
 (20.7456328, -103.4172233),
 (20.7220746, -103.3919446),
 (20.6874171, -103.2770813),
 (20.6305973, -103.3752093),
 (20.7160767, -103.3150292),
 (20.7037164, -103.29088),
 (20.6794019, -103.274831),
 (20.6879047, -103.3

In [13]:
df = pd.DataFrame(koordinaten, columns =['latitude', 'longitude'])
df

latitude   longitude
0    20.679439 -103.269892
1    20.696684 -103.315000
2    20.716077 -103.315029
3    20.657974 -103.366088
4    20.647878 -103.378214
..         ...         ...
241  20.671271 -103.333563
242  20.624525 -103.412610
243  20.769739 -103.446289
244  20.619582 -103.380122
245  20.643346 -103.324832

[246 rows x 2 columns]

In [14]:
df1 = pd.DataFrame({'demand' : df.groupby( [ "latitude", "longitude"] ).size()}).reset_index()
df1

latitude   longitude  demand
0    20.607961 -103.345722       1
1    20.608084 -103.346261       1
2    20.613829 -103.348514       2
3    20.613957 -103.342518       2
4    20.614029 -103.351272       1
..         ...         ...     ...
179  20.745633 -103.417223       2
180  20.760553 -103.338346       1
181  20.769739 -103.446289       2
182  20.788996 -103.442492       1
183  20.808935 -103.457485       1

[184 rows x 3 columns]

In [15]:
# customer count ('0' is depot) 
customer_count = len(df1)

# the number of vehicle
vehicle_count = 26

# the capacity of vehicle
vehicle_capacity = 30

# set depot latitude and longitude
depot_latitude = 20.6489176 
depot_longitude = -103.3409776

In [16]:
df1.loc[-1] = [depot_latitude, depot_longitude, 0]  # adding a row
df1.index = df1.index + 1  # shifting index
df1.sort_index(inplace=True) 

In [32]:
# function for calculating distance between two pins
def _distance_calculator(_df):
    
    _distance_result = np.zeros((len(_df),len(_df)))
    _df['latitude-longitude'] = '0'
    for i in range(len(_df)):
        _df['latitude-longitude'].iloc[i] = str(_df.latitude[i]) + ',' + str(_df.longitude[i])
    
    for i in range(len(_df)):
        for j in range(len(_df)):
            
            # calculate distance of all pairs
            _google_maps_api_result = googlemaps.directions(_df['latitude-longitude'].iloc[i],
                                                            _df['latitude-longitude'].iloc[j],
                                                            mode = 'driving')
            # append distance to result list
            _distance_result[i][j] = _google_maps_api_result[0]['legs'][0]['distance']['value']
    
    return _distance_result

In [34]:
distance = _distance_calculator(df1)

In [35]:
distance

array([[    0.,  8802.,  8751., ..., 29041., 32024., 35731.],
       [ 7058.,     0.,    51., ..., 28453., 31436., 35143.],
       [ 7007.,    51.,     0., ..., 28402., 31385., 35092.],
       ...,
       [28953., 31288., 31237., ...,     0.,  2829.,  6706.],
       [31956., 34291., 34240., ...,  3390.,     0.,  7031.],
       [33442., 35777., 35726., ...,  6040.,  5754.,     0.]])

In [37]:
from numpy import savetxt

In [38]:
savetxt('distances_zmg.csv', distance, delimiter=',')

In [39]:
from ortools.constraint_solver import routing_enums_pb2
from ortools.constraint_solver import pywrapcp

In [40]:
def create_data_model():
    """Stores the data for the problem."""
    data = {}
    data['distance_matrix'] = distance
    data['demands'] = df1['demand']
    data['vehicle_capacities'] = [30, 30, 30, 30, 30, 
                                 30, 30, 30, 30, 30,
                                 30]
    data['num_vehicles'] = 11
    data['depot'] = 0
    return data


def print_solution(data, manager, routing, solution):
    """Prints solution on console."""
    print(f'Objective: {solution.ObjectiveValue()}')
    total_distance = 0
    total_load = 0
    for vehicle_id in range(data['num_vehicles']):
        index = routing.Start(vehicle_id)
        plan_output = 'Route for vehicle {}:\n'.format(vehicle_id)
        route_distance = 0
        route_load = 0
        while not routing.IsEnd(index):
            node_index = manager.IndexToNode(index)
            route_load += data['demands'][node_index]
            plan_output += ' {0} Load({1}) -> '.format(node_index, route_load)
            previous_index = index
            index = solution.Value(routing.NextVar(index))
            route_distance += routing.GetArcCostForVehicle(
                previous_index, index, vehicle_id)
        plan_output += ' {0} Load({1})\n'.format(manager.IndexToNode(index),
                                                 route_load)
        plan_output += 'Distance of the route: {}m\n'.format(route_distance)
        plan_output += 'Load of the route: {}\n'.format(route_load)
        print(plan_output)
        total_distance += route_distance
        total_load += route_load
    print('Total distance of all routes: {}m'.format(total_distance))
    print('Total load of all routes: {}'.format(total_load))


def main():
    """Solve the CVRP problem."""
    # Instantiate the data problem.
    data = create_data_model()

    # Create the routing index manager.
    manager = pywrapcp.RoutingIndexManager(len(data['distance_matrix']),
                                           data['num_vehicles'], data['depot'])

    # Create Routing Model.
    routing = pywrapcp.RoutingModel(manager)


    # Create and register a transit callback.
    def distance_callback(from_index, to_index):
        """Returns the distance between the two nodes."""
        # Convert from routing variable Index to distance matrix NodeIndex.
        from_node = manager.IndexToNode(from_index)
        to_node = manager.IndexToNode(to_index)
        return data['distance_matrix'][from_node][to_node]

    transit_callback_index = routing.RegisterTransitCallback(distance_callback)

    # Define cost of each arc.
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)


    # Add Capacity constraint.
    def demand_callback(from_index):
        """Returns the demand of the node."""
        # Convert from routing variable Index to demands NodeIndex.
        from_node = manager.IndexToNode(from_index)
        return data['demands'][from_node]

    demand_callback_index = routing.RegisterUnaryTransitCallback(
        demand_callback)
    routing.AddDimensionWithVehicleCapacity(
        demand_callback_index,
        0,  # null capacity slack
        data['vehicle_capacities'],  # vehicle maximum capacities
        True,  # start cumul to zero
        'Capacity')

    # Setting first solution heuristic.
    search_parameters = pywrapcp.DefaultRoutingSearchParameters()
    search_parameters.first_solution_strategy = (
        routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC)
    search_parameters.local_search_metaheuristic = (
        routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH)
    search_parameters.time_limit.FromSeconds(1)

    # Solve the problem.
    solution = routing.SolveWithParameters(search_parameters)

    # Print solution on console.
    if solution:
        print_solution(data, manager, routing, solution)


if __name__ == '__main__':
    main()

Objective: 422948
Route for vehicle 0:
 0 Load(0.0) ->  0 Load(0.0)
Distance of the route: 0m
Load of the route: 0.0

Route for vehicle 1:
 0 Load(0.0) ->  0 Load(0.0)
Distance of the route: 0m
Load of the route: 0.0

Route for vehicle 2:
 0 Load(0.0) ->  70 Load(1.0) ->  34 Load(2.0) ->  9 Load(3.0) ->  10 Load(4.0) ->  39 Load(5.0) ->  60 Load(6.0) ->  125 Load(7.0) ->  177 Load(8.0) ->  180 Load(10.0) ->  182 Load(12.0) ->  183 Load(13.0) ->  184 Load(14.0) ->  134 Load(15.0) ->  96 Load(16.0) ->  73 Load(17.0) ->  53 Load(25.0) ->  13 Load(26.0) ->  32 Load(27.0) ->  27 Load(28.0) ->  55 Load(29.0) ->  57 Load(30.0) ->  0 Load(30.0)
Distance of the route: 93967m
Load of the route: 30.0

Route for vehicle 3:
 0 Load(0.0) ->  98 Load(1.0) ->  101 Load(2.0) ->  120 Load(3.0) ->  128 Load(4.0) ->  139 Load(5.0) ->  149 Load(6.0) ->  114 Load(8.0) ->  118 Load(9.0) ->  107 Load(10.0) ->  61 Load(11.0) ->  28 Load(12.0) ->  17 Load(15.0) ->  16 Load(18.0) ->  25 Load(20.0) ->  38 Load(21

Objective: 422948
Route for vehicle 0:
 0 Load(0.0) ->  0 Load(0.0)
Distance of the route: 0m
Load of the route: 0.0

Route for vehicle 1:
 0 Load(0.0) ->  0 Load(0.0)
Distance of the route: 0m
Load of the route: 0.0

Route for vehicle 2:
 0 Load(0.0) ->  70 Load(1.0) ->  34 Load(2.0) ->  9 Load(3.0) ->  10 Load(4.0) ->  39 Load(5.0) ->  60 Load(6.0) ->  125 Load(7.0) ->  177 Load(8.0) ->  180 Load(10.0) ->  182 Load(12.0) ->  183 Load(13.0) ->  184 Load(14.0) ->  134 Load(15.0) ->  96 Load(16.0) ->  73 Load(17.0) ->  53 Load(25.0) ->  13 Load(26.0) ->  32 Load(27.0) ->  27 Load(28.0) ->  55 Load(29.0) ->  57 Load(30.0) ->  0 Load(30.0)
Distance of the route: 93967m
Load of the route: 30.0

Route for vehicle 3:
 0 Load(0.0) ->  98 Load(1.0) ->  101 Load(2.0) ->  120 Load(3.0) ->  128 Load(4.0) ->  139 Load(5.0) ->  149 Load(6.0) ->  114 Load(8.0) ->  118 Load(9.0) ->  107 Load(10.0) ->  61 Load(11.0) ->  28 Load(12.0) ->  17 Load(15.0) ->  16 Load(18.0) ->  25 Load(20.0) ->  38 Load(21.0) ->  41 Load(22.0) ->  40 Load(23.0) ->  22 Load(24.0) ->  12 Load(25.0) ->  7 Load(26.0) ->  8 Load(27.0) ->  11 Load(29.0) ->  50 Load(30.0) ->  0 Load(30.0)
Distance of the route: 53804m
Load of the route: 30.0

Route for vehicle 4:
 0 Load(0.0) ->  43 Load(2.0) ->  54 Load(3.0) ->  156 Load(10.0) ->  174 Load(12.0) ->  181 Load(13.0) ->  178 Load(14.0) ->  173 Load(16.0) ->  169 Load(17.0) ->  170 Load(18.0) ->  172 Load(19.0) ->  175 Load(20.0) ->  179 Load(21.0) ->  176 Load(22.0) ->  163 Load(24.0) ->  160 Load(26.0) ->  159 Load(28.0) ->  171 Load(29.0) ->  167 Load(30.0) ->  0 Load(30.0)
Distance of the route: 58693m
Load of the route: 30.0

Route for vehicle 5:
 0 Load(0.0) ->  74 Load(1.0) ->  51 Load(2.0) ->  45 Load(4.0) ->  49 Load(5.0) ->  71 Load(6.0) ->  87 Load(7.0) ->  56 Load(8.0) ->  46 Load(9.0) ->  19 Load(10.0) ->  23 Load(11.0) ->  20 Load(12.0) ->  37 Load(13.0) ->  0 Load(13.0)
Distance of the route: 35107m
Load of the route: 13.0

Route for vehicle 6:
 0 Load(0.0) ->  108 Load(3.0) ->  121 Load(4.0) ->  132 Load(5.0) ->  130 Load(7.0) ->  147 Load(9.0) ->  162 Load(10.0) ->  161 Load(11.0) ->  165 Load(12.0) ->  166 Load(13.0) ->  157 Load(14.0) ->  154 Load(16.0) ->  153 Load(17.0) ->  155 Load(18.0) ->  158 Load(19.0) ->  164 Load(20.0) ->  168 Load(21.0) ->  102 Load(23.0) ->  97 Load(25.0) ->  0 Load(25.0)
Distance of the route: 41775m
Load of the route: 25.0

Route for vehicle 7:
 0 Load(0.0) ->  62 Load(1.0) ->  35 Load(2.0) ->  21 Load(3.0) ->  18 Load(4.0) ->  14 Load(5.0) ->  5 Load(6.0) ->  3 Load(8.0) ->  6 Load(9.0) ->  4 Load(11.0) ->  2 Load(12.0) ->  1 Load(13.0) ->  15 Load(14.0) ->  26 Load(15.0) ->  24 Load(16.0) ->  29 Load(18.0) ->  36 Load(20.0) ->  31 Load(22.0) ->  33 Load(23.0) ->  30 Load(24.0) ->  47 Load(25.0) ->  48 Load(26.0) ->  42 Load(28.0) ->  0 Load(28.0)
Distance of the route: 32145m
Load of the route: 28.0

Route for vehicle 8:
 0 Load(0.0) ->  76 Load(1.0) ->  94 Load(2.0) ->  126 Load(3.0) ->  122 Load(4.0) ->  129 Load(5.0) ->  131 Load(6.0) ->  136 Load(8.0) ->  133 Load(10.0) ->  143 Load(11.0) ->  137 Load(12.0) ->  148 Load(18.0) ->  142 Load(19.0) ->  105 Load(21.0) ->  116 Load(22.0) ->  113 Load(23.0) ->  106 Load(25.0) ->  88 Load(27.0) ->  83 Load(29.0) ->  75 Load(30.0) ->  0 Load(30.0)
Distance of the route: 36716m
Load of the route: 30.0

Route for vehicle 9:
 0 Load(0.0) ->  44 Load(1.0) ->  52 Load(2.0) ->  58 Load(3.0) ->  66 Load(4.0) ->  68 Load(5.0) ->  65 Load(6.0) ->  59 Load(7.0) ->  85 Load(9.0) ->  86 Load(10.0) ->  79 Load(11.0) ->  80 Load(12.0) ->  82 Load(13.0) ->  91 Load(14.0) ->  95 Load(15.0) ->  99 Load(17.0) ->  124 Load(18.0) ->  140 Load(21.0) ->  145 Load(22.0) ->  146 Load(23.0) ->  151 Load(24.0) ->  150 Load(25.0) ->  152 Load(26.0) ->  141 Load(27.0) ->  144 Load(28.0) ->  138 Load(29.0) ->  135 Load(30.0) ->  0 Load(30.0)
Distance of the route: 38119m
Load of the route: 30.0

Route for vehicle 10:
 0 Load(0.0) ->  63 Load(1.0) ->  67 Load(2.0) ->  84 Load(3.0) ->  90 Load(4.0) ->  92 Load(5.0) ->  93 Load(6.0) ->  81 Load(7.0) ->  72 Load(8.0) ->  64 Load(10.0) ->  69 Load(11.0) ->  77 Load(12.0) ->  89 Load(13.0) ->  100 Load(14.0) ->  104 Load(16.0) ->  109 Load(17.0) ->  110 Load(19.0) ->  111 Load(20.0) ->  103 Load(21.0) ->  112 Load(22.0) ->  127 Load(24.0) ->  123 Load(25.0) ->  119 Load(26.0) ->  117 Load(27.0) ->  115 Load(29.0) ->  78 Load(30.0) ->  0 Load(30.0)
Distance of the route: 32622m
Load of the route: 30.0

Total distance of all routes: 422948m
Total load of all routes: 246.0